moreh norm is failing on L2 nightly

https://github.com/tenstorrent/tt-metal/actions/runs/27956266846/job/82730140775#step:6:19512
```
FAILED tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_norm.py::test_moreh_norm_rank_1_dim_0_keepdim_false_output_modes[is_linalg_vector_norm=False-provide-output-p=0.0] - RuntimeError: output with shape [] doesn't match the broadcast shape [1]
```

--- a/tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_norm.py
+++ b/tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_norm.py
@@ -59,6 +59,16 @@ def torch_norm(
     """
     if is_linalg_vector_norm:
         torch_output = torch.linalg.vector_norm(torch_input, ord=p, dim=dim, keepdim=keepdim)
+    elif not keepdim and dim is not None and dim != []:
+        # The deprecated torch.norm mishandles keepdim=False for some orders (e.g. p=0)
+        # when the reduction collapses to a scalar, raising
+        # "output with shape [] doesn't match the broadcast shape [1]". keepdim=True is
+        # unaffected, so reduce with keepdim and squeeze the reduced dims ourselves.
+        torch_output = torch.norm(torch_input, p=p, dim=dim, keepdim=True)
+        reduce_dims = [dim] if isinstance(dim, int) else list(dim)
+        rank = torch_input.dim()
+        for d in sorted((d % rank for d in reduce_dims), reverse=True):
+            torch_output = torch_output.squeeze(d)
     else:
         torch_output = torch.norm(torch_input, p=p, dim=dim, keepdim=keepdim)
     torch_input_grad = None
@@ -240,6 +250,11 @@ def run_moreh_norm_output_mode(
 
     actual_output = ttnn.operations.moreh.norm(ttnn_input, **kwargs)
     actual_output = ttnn.to_torch(actual_output)
+
+    # For a rank-1 input reduced with keepdim=False the torch reference is rank-0 ([]) while
+    # the ttnn output keeps a trailing dim ([1]); they are the same scalar. Match shapes so the
+    # comparison doesn't trip torch's "shape [] doesn't match broadcast shape [1]" error.
+    expected_output = expected_output.reshape(actual_output.shape)
 
     passing, out = comp_allclose(expected_output, actual_output, rtol=rtol, atol=atol)
     logger.info(f"output's {out}")

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp
@@ -19,23 +19,12 @@ Tensor moreh_norm(
     const std::optional<Tensor>& output,
     const std::optional<MemoryConfig>& memory_config,
     const std::optional<DeviceComputeKernelConfig>& compute_kernel_config) {
+    // get_dim() normalizes dim into a concrete list of dims, handling every form uniformly:
+    // nullopt (Python dim=None) and an empty list both expand to the full dim range, a single
+    // int becomes {d}, and a list is taken as-is. Single-dim cases are then handled by the
+    // dims.size() == 1 branch below, so dim is never accessed via dim.value() directly.
     auto dims = operations::get_dim(dim, input.logical_shape().rank());
     auto INF = std::numeric_limits<float>::infinity();
-    // Guard dim.value(): when dim is nullopt (e.g. Python dim=None for a global norm), accessing
-    // it would throw std::bad_optional_access. In that case dims already holds the full dim range
-    // from get_dim(), so fall through to the multi-dim handling below.
-    if (dim.has_value()) {
-        if (auto* single_dim = std::get_if<int64_t>(&dim.value())) {
-            if (p == 0.0 || p == INF || p == -INF) {
-                return ttnn::prim::moreh_norm(
-                    input, p, *single_dim, keepdim, output, memory_config, compute_kernel_config);
-            }
-            auto tmp_output = ttnn::moreh_abs_pow(input, p, std::nullopt, memory_config, compute_kernel_config);
-            tmp_output =
-                ttnn::moreh_sum(tmp_output, *single_dim, keepdim, std::nullopt, memory_config, compute_kernel_config);
-            return ttnn::moreh_abs_pow(tmp_output, 1.0f / p, output, memory_config, compute_kernel_config);
-        }
-    }
 
     if (dims.size() == 1) {
         if (p == 0.0 || p == INF || p == -INF) {

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp
@@ -21,14 +21,20 @@ Tensor moreh_norm(
     const std::optional<DeviceComputeKernelConfig>& compute_kernel_config) {
     auto dims = operations::get_dim(dim, input.logical_shape().rank());
     auto INF = std::numeric_limits<float>::infinity();
-    if (auto* single_dim = std::get_if<int64_t>(&dim.value())) {
-        if (p == 0.0 || p == INF || p == -INF) {
-            return ttnn::prim::moreh_norm(input, p, *single_dim, keepdim, output, memory_config, compute_kernel_config);
+    // Guard dim.value(): when dim is nullopt (e.g. Python dim=None for a global norm), accessing
+    // it would throw std::bad_optional_access. In that case dims already holds the full dim range
+    // from get_dim(), so fall through to the multi-dim handling below.
+    if (dim.has_value()) {
+        if (auto* single_dim = std::get_if<int64_t>(&dim.value())) {
+            if (p == 0.0 || p == INF || p == -INF) {
+                return ttnn::prim::moreh_norm(
+                    input, p, *single_dim, keepdim, output, memory_config, compute_kernel_config);
+            }
+            auto tmp_output = ttnn::moreh_abs_pow(input, p, std::nullopt, memory_config, compute_kernel_config);
+            tmp_output =
+                ttnn::moreh_sum(tmp_output, *single_dim, keepdim, std::nullopt, memory_config, compute_kernel_config);
+            return ttnn::moreh_abs_pow(tmp_output, 1.0f / p, output, memory_config, compute_kernel_config);
         }
-        auto tmp_output = ttnn::moreh_abs_pow(input, p, std::nullopt, memory_config, compute_kernel_config);
-        tmp_output =
-            ttnn::moreh_sum(tmp_output, *single_dim, keepdim, std::nullopt, memory_config, compute_kernel_config);
-        return ttnn::moreh_abs_pow(tmp_output, 1.0f / p, output, memory_config, compute_kernel_config);
     }
 
     if (dims.size() == 1) {

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp
@@ -14,7 +14,7 @@ namespace ttnn {
 Tensor moreh_norm(
     const Tensor& input,
     float p,
-    std::optional<std::variant<int64_t, ttnn::SmallVector<int64_t>>> dim,
+    const std::optional<std::variant<int64_t, ttnn::SmallVector<int64_t>>>& dim,
     bool keepdim,
     const std::optional<Tensor>& output,
     const std::optional<MemoryConfig>& memory_config,

--- a/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.hpp
+++ b/ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.hpp
@@ -12,7 +12,7 @@ namespace ttnn {
 Tensor moreh_norm(
     const Tensor& input,
     float p,
-    std::optional<std::variant<int64_t, ttnn::SmallVector<int64_t>>> dim = std::nullopt,
+    const std::optional<std::variant<int64_t, ttnn::SmallVector<int64_t>>>& dim = std::nullopt,
     bool keepdim = false,
     const std::optional<Tensor>& output = std::nullopt,
     const std::optional<MemoryConfig>& memory_config = std::nullopt,

